lib/CsvImportAuditer.class.php

```diff
@@ -0,0 +1,308 @@
+    {
+        $this->ormClasses = $classes;
+    }
+
+    public function setSourceName($sourceName): void
+    {
+        $this->sourceName = $sourceName;
+    }
+
+    public function getSourceName(): string
+    {
+        return $this->sourceName ?? '';
+    }
+
+    public function setTargetName($targetName): void
+    {
+        if (empty($targetName)) {
+            throw new ValueError("Target name can't be blank.");
+        }
+
+        $this->targetName = $targetName;
+    }
+
+    public function getTargetName(): string
+    {
+        return $this->targetName;
+    }
+
+    public function setFilename($filename): void
+    {
+        $this->filename = $this->validateFilename($filename);
+
+        if (empty($this->getSourceName())) {
+            $this->setSourceName(basename($filename));
+        }
+    }
+
+    public function getFilename(): string
+    {
+        return $this->filename;
+    }
+
+    public function validateFilename($filename): string
+    {
+        if (empty($filename)) {
+            throw new sfException('Please specify a filename for import');
+        }
+
+        if (!file_exists($filename)) {
+            throw new sfException("Can not find file {$filename}");
+        }
+
+        if (!is_readable($filename)) {
+            throw new sfException("Can not read {$filename}");
+        }
+
+        return $filename;
+    }
+
+    public function setOptions(array $options = null): void
+    {
+        if (empty($options)) {
+            return;
+        }
+
+        foreach ($options as $name => $val) {
+            $this->setOption($name, $val);
+        }
+    }
+
+    public function getOptions(): array
+    {
+        return $this->options;
+    }
+
+    public function setOption(string $name, $value): void
+    {
+        $this->options[$name] = $value;
+    }
+
+    public function getOption(string $name)
+    {
+        return $this->options[$name] ?? null;
+    }
+
+    public function countRowsAudited(): int
+    {
+        return $this->rowsAudited;
+    }
+
+    public function countRowsTotal(): int
+    {
+        return $this->rowsTotal;
+    }
+
+    public function doAudit($filename = null): void
+    {
+        if (null !== $filename) {
+            $this->setFilename($filename);
+        }
+
+        $records = $this->loadCsvData($this->filename);
+
+        foreach ($records as $record) {
+            try {
+                $this->processRow($record);
+            } catch (UnexpectedValueException $e) {
+                $this->logError(sprintf(
+                    'Warning! skipped row [%u/%u]: %s',
+                    $this->rowsAudited,
+                    $this->rowsTotal,
+                    $e->getMessage()
+                ));
+
+                continue;
+            }
+
+            ++$this->rowsAudited;
+            $this->log($this->progressUpdate($this->rowsAudited));
+        }
+
+        if (!empty($this->missingIds)) {
+            $this->log('');
```

Comment:
Is logging an empty string here necessary?

Reply:
I was hacking in a blank line but I'll see if I can do that but explicitly adding one instead.